### PR TITLE
✨ URL parsing service

### PR DESCRIPTION
**Pull request by jridgewell**
Adds a `AmpDoc` service for parsing URLs.

Part of https://github.com/ampproject/amphtml/issues/14753

/cc @jpettitt

### CODE REVIEW

File: extensions/amp-a4a/0.1/amp-ad-templates.js [hunk 1]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-import {LRUCache} from '../../../src/utils/lru-cache';
+import {LruCache} from '../../../src/utils/lru-cache';
```
Comment (kristoferbaxter) [inline extensions/amp-a4a/0.1/amp-ad-templates.js]: Looks unrelated. Accidental inclusion?

File: src/utils/lru-cache.js [hunk 2]
```diff
@@ -16,49 +16,76 @@
...
-      dev().warn(TAG, 'Trimming template cache');
-      // Evict oldest entry to ensure memory usage is minimized.
-      cacheKeys.sort((a, b) => this.cache_[b].access - this.cache_[a].access);
-      delete this.cache_[cacheKeys[cacheKeys.length - 1]];
+    this.cache_[id] = {payload, access: this.access_};
+    this.size_++;
+    this.evict_();
+  }
+
+  evict_() {
```
Comment (kristoferbaxter) [inline src/utils/lru-cache.js]: JSDoc might be a good idea here.

**Comment by jridgewell** [comment]
Ping @choumx

File: src/utils/lru-cache.js [hunk 3]
```diff
@@ -16,49 +16,89 @@
...
 
-    /** @private {!Object<(number|string), !Cacheable>} */
+    /** @private {number} */
+    this.size_ = 0;
+
+    /**
+     * An incrementing counter to define the last access.
+     * @private {number}
+     */
+    this.access_ = 0;
```
Comment (choumx) [inline src/utils/lru-cache.js]: TIL that number overflow is not an issue.

File: src/utils/lru-cache.js [hunk 4]
```diff
@@ -16,49 +16,89 @@
...
+    for (const key in cache) {
+      const {access} = cache[key];
+      if (access < oldest) {
+        oldest = access;
+        oldestKey = key;
+      }
+    }
+
+    if (oldestKey !== undefined) {
+      delete cache[oldestKey];
```
Comment (choumx) [inline src/utils/lru-cache.js]: Shouldn't decrementing `this.size_` be moved here to avoid skew?

File: src/utils/lru-cache.js [hunk 5]
```diff
@@ -16,49 +16,89 @@
...
-      delete this.cache_[cacheKeys[cacheKeys.length - 1]];
+  evict_() {
+    if (this.size_ <= this.capacity_) {
+      return;
+    }
+    this.size_--;
+
+    dev().warn(TAG, 'Trimming LRU cache');
+    const cache = this.cache_;
+    let oldest = this.access_ + 1;
```
Comment (choumx) [inline src/utils/lru-cache.js]: Nit: Start with `0` and increase to avoid `MAX_VALUE + 1 == MAX_VALUE` problem.

File: test/functional/utils/test-lru-cache.js [hunk 6]
```diff
@@ -14,28 +14,34 @@
...
   beforeEach(() => {
-    cache = new LRUCache(5);
+    cache = new LruCache(5);
     for (let i = 0; i < 5; i++) {
-      cache.cache_[i] = {payload: i, access: i};
+      cache.put(i, i);
     }
   });
 
   it('should create a protype-less object for caching', () => {
```
Comment (choumx) [inline test/functional/utils/test-lru-cache.js]: "prototype-less"

File: src/utils/lru-cache.js [hunk 7]
```diff
@@ -16,49 +16,89 @@
...
 
   /**
-   * @param {number|string} id
-   * @param {*} payload The payload to cache.
+   * @param {number|string} key
+   * @param {T} payload The payload to cache.
+   */
+  put(key, payload) {
+    this.cache_[key] = {payload, access: this.access_};
+    this.size_++;
```
Comment (choumx) [inline src/utils/lru-cache.js]: The size doesn't increase if the key already exists, right?

File: src/service/url-impl.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+  /**
+   * Parses the URL in the context of the current document.
+   *
+   * @param {string} url
+   * @param {boolean=} opt_nocache
+   * @return {!Location}
+   */
+  parse(url, opt_nocache) {
+    return parseUrlWithA(this.anchor_, url, opt_nocache ? null : this.cache_);
```
Comment (choumx) [inline src/service/url-impl.js]: Did we capture all of the "heavy" parts of URL parsing in this class? I'm wondering if we should move other URL parsing functions here. Or are the remaining methods pretty light?

File: src/runtime.js [hunk 9]
```diff
@@ -105,6 +106,7 @@ export function installRuntimeServices(global) {
  * @param {!Object<string, string>=} opt_initParams
  */
 export function installAmpdocServices(ampdoc, opt_initParams) {
+  installUrlForDoc(ampdoc);
```
Comment (choumx) [inline src/runtime.js]: Do we already do some URL parsing on startup? Wondering if this would impact initial load.

Comment (jridgewell) [reply src/runtime.js]: Yah, in viewer, doc-info, preconnect, etc. It's all over the place.

File: src/service/url-impl.js (hunk 8 above)
Comment (jridgewell) [reply src/service/url-impl.js]: There will be several more methods getting added. For now, we can take care of the largest usecase with just `parseUrl`.

File: src/utils/lru-cache.js (hunk 3 above)
Comment (jridgewell) [reply src/utils/lru-cache.js]: Yah, `2 ** 53` is a very large number. For reference, we can do 1,000 accesses every millisecond for 285 years before we hit it.

File: src/utils/lru-cache.js (hunk 7 above)
Comment (jridgewell) [reply src/utils/lru-cache.js]: Hm, this was implicitly handled before. I'll add a check and tests.

File: src/utils/lru-cache.js (hunk 5 above)
Comment (jridgewell) [reply src/utils/lru-cache.js]: > Start with 0 and increase

How would I code that?

> to avoid MAX_VALUE + 1 == MAX_VALUE problem

See https://github.com/ampproject/amphtml/pull/15414/files#r189728551.

File: src/utils/lru-cache.js (hunk 4 above)
Comment (jridgewell) [reply src/utils/lru-cache.js]: This is to satisfy Closure. There's no actual way to not find a `oldestKey` in real code, due to the `this.size_ <= this.capacity_`.

Either way, I'll move it in here.

File: src/utils/lru-cache.js (hunk 3 above)
Comment (choumx) [reply src/utils/lru-cache.js]: I meant in JS in general, but +1 for napkin math!
